components/supervisor/pkg/supervisor/tasks_test.go

```diff
@@ -0,0 +1,152 @@
+	}{
+		{
+			Desc:     "headless prebuild should finish without tasks",
+			Headless: true,
+			Source:   api.WorkspaceInitFromOther,
+
+			ExpectedReporter: testHeadlessTaskProgressReporter{
+				Done:    true,
+				Success: true,
+			},
+		},
+		{
+			Desc:        "headless prebuild should finish without init tasks",
+			Headless:    true,
+			Source:      api.WorkspaceInitFromOther,
+			GitpodTasks: &[]TaskConfig{{Command: &skipCommand}},
+
+			ExpectedReporter: testHeadlessTaskProgressReporter{
+				Done:    true,
+				Success: true,
+			},
+		},
+		{
+			Desc:        "headless prebuild should finish with successful init tasks",
+			Headless:    true,
+			Source:      api.WorkspaceInitFromOther,
+			GitpodTasks: &[]TaskConfig{{Init: &skipCommand}, {Init: &skipCommand}},
+
+			ExpectedReporter: testHeadlessTaskProgressReporter{
+				Done:    true,
+				Success: true,
+			},
+		},
+		{
+			Desc:        "headless prebuild should finish with failed init tasks",
+			Headless:    true,
+			Source:      api.WorkspaceInitFromOther,
+			GitpodTasks: &[]TaskConfig{{Init: &failCommand}, {Init: &failCommand}},
+
+			ExpectedReporter: testHeadlessTaskProgressReporter{
+				Done:    true,
+				Success: false,
+			},
+		},
+		{
+			Desc:        "headless prebuild should finish with at least one failed init tasks (first)",
+			Headless:    true,
+			Source:      api.WorkspaceInitFromOther,
+			GitpodTasks: &[]TaskConfig{{Init: &failCommand}, {Init: &skipCommand}},
+
+			ExpectedReporter: testHeadlessTaskProgressReporter{
+				Done:    true,
+				Success: false,
+			},
+		},
+		{
+			Desc:        "headless prebuild should finish with at least one failed init tasks (second)",
+			Headless:    true,
+			Source:      api.WorkspaceInitFromOther,
+			GitpodTasks: &[]TaskConfig{{Init: &skipCommand}, {Init: &failCommand}},
+
+			ExpectedReporter: testHeadlessTaskProgressReporter{
+				Done:    true,
+				Success: false,
+			},
+		},
+	}
+	for _, test := range tests {
+		t.Run(test.Desc, func(t *testing.T) {
+			storeLocation, err := ioutil.TempDir("", "tasktest")
+			if err != nil {
+				t.Fatal(err)
+			}
+			defer os.RemoveAll(storeLocation)
+
+			gitpodTasks := ""
+			if test.GitpodTasks != nil {
+				result, err := json.Marshal(test.GitpodTasks)
+				if err != nil {
+					t.Fatal(err)
+				}
+				gitpodTasks = string(result)
+			}
+
+			var (
+				terminalService = terminal.NewMuxTerminalService(terminal.NewMux())
+				contentState    = NewInMemoryContentState("")
+				reporter        = testHeadlessTaskProgressReporter{}
+				taskManager     = newTasksManager(&Config{
+					WorkspaceConfig: WorkspaceConfig{
+						GitpodTasks:    gitpodTasks,
+						GitpodHeadless: strconv.FormatBool(test.Headless),
+					},
+				}, terminalService, contentState, &reporter)
+			)
+			taskManager.storeLocation = storeLocation
+			contentState.MarkContentReady(test.Source)
+			var wg sync.WaitGroup
+			wg.Add(1)
+			taskManager.Run(context.Background(), &wg)
```

Comment:
there's no `wg.Wait()` ... is that intentional?

Reply:
run is blocking, but i can add it just to be sure

Reply:
makes sense. Either `wg.Wait()` or a comment would be handy :)